frontend/apps/web/components/jobs/NosqlTable/NosqlTable.tsx

```diff
@@ -116,6 +122,50 @@ export default function NosqlTable(props: Props): ReactElement {
 
   const collections = Array.from(Object.keys(schema));
 
+  // useMemo ensures that we don't recreate the set unless the data changes
+  const keySet = useMemo(() => {
+    const set = new Set<string>();
+    data.forEach((item: JobMappingFormValues) => {
+      set.add(`${item.schema}.${item.table}.${item.column}`);
+    });
+    return set;
+  }, [data]);
+
+  // useCallback ensures that we only re-run the function if the keySet changes
+  // we pass in an optional curr value so that we don't throw errors when searching for duplicates and the existing key is there
+  // for ex. if the initial value is "email" and we change it to "emaila" and then back to "email", it throws an error because the key set is made from all of the rows of the data which includes the original value, so we delete it if currValue is valid so we don't throw that error and allow the user to just save their current value without changing it
+  const isDuplicateKey = useCallback(
+    (newValue: string, schema: string, table: string, currValue?: string) => {
+      const key = `${schema}.${table}.${newValue}`;
+      if (currValue) {
+        keySet.delete(`${schema}.${table}.${currValue}`);
+      }
```

Comment:
This is a little scary. something called `isDuplicateKey` I would not expect to modify the key set.
I would expect the `isDuplicateKey` to be really dumb and basically just be doing this:

```
return keySet.has(`${schema}.${table}.${newValue}`);
```
and nothing else.